src/Training.AirBnb.Clone.Backend/Backend Project/Domain/Services/AddressService.cs

```diff
@@ -0,0 +1,153 @@
+        }
+
+        public IQueryable<Address> Get(Expression<Func<Address, bool>> predicate)
+        {
+            return GetUndeletedAddresses().Where(predicate.Compile()).AsQueryable();
+        }
+
+        public ValueTask<ICollection<Address>> GetAsync(IEnumerable<Guid> ids, CancellationToken cancellationToken = default)
+        {
+            var addresses = GetUndeletedAddresses().
+                Where(address => ids.Contains(address.Id));
+            return new ValueTask<ICollection<Address>>(addresses.ToList());
+        }
+
+        public ValueTask<Address> GetByIdAsync(Guid id, CancellationToken cancellationToken = default)
+        {
+            return new ValueTask<Address>(GetUndeletedAddresses().
+                FirstOrDefault(address => address.Id == id) ??
+                throw new AddressNotFoundException("Address not found"));
+        }
+
+        public async ValueTask<Address> UpdateAsync(Address address, bool saveChanges = true, CancellationToken cancellationToken = default)
+        {
+            var updatedAddress = GetUndeletedAddresses().
+                FirstOrDefault(updateAddress => updateAddress.Id.Equals(address.Id));
+
+            if (updatedAddress is null)
+                throw new AddressNotFoundException("Address not found!");
+
+            if (!IsValidAddressLines(address.AddressLine1))
+                throw new AddressFormatException("Invalid province!");
+            if (!IsValidZipCode(address.ZipCode))
+                throw new AddressFormatException("Invalid zipCode!");
+            if (!(await IsUniqueCountry(address)))
+                throw new AddressFormatException();
+
+            updatedAddress.CityId = address.CityId;
+            updatedAddress.CountryId = address.CountryId;
+            updatedAddress.AddressLine1 = address.AddressLine1;
+            updatedAddress.AddressLine2 = address.AddressLine2;
+            updatedAddress.AddressLine3 = address.AddressLine3;
+            updatedAddress.AddressLine4 = address.AddressLine4;
+            updatedAddress.Province = address.Province;
+            updatedAddress.ZipCode = address.ZipCode;
+            updatedAddress.ModifiedDate = DateTimeOffset.UtcNow;
+
+            if(saveChanges)
+                await _appDataContext.SaveChangesAsync();
+            
+            return updatedAddress;
+
+        }
+
+        private bool IsValidAddressLines(string addressLine)
+        {
+            if (!string.IsNullOrWhiteSpace(addressLine))
+                return true;
+            else
+                return false;
+        }
+
+        private bool IsValidZipCode(string? zipCode)
+        {
+            if(zipCode is null)
+                return true;
+            for (int index = 0; index < zipCode?.Length; index++)
+                if (!char.IsNumber(zipCode[index]))
+                    return false;
+            return true;
+        }
+
+        private async ValueTask<bool> IsUniqueCountry(Address adress)
+        {
```

Comment:
Shu method nomini to'g'irlash kerak, chunki bu method unique likka tekshirmaydi, city country ga tegishli ekanligini tekshiradi, shunga mos nom qo'yish kerak.